src/rgw/rgw_sal.cc

```diff
@@ -83,161 +72,61 @@ RGWObjState::RGWObjState(const RGWObjState& rhs) : obj (rhs.obj) {
 
 rgw::sal::Store* StoreManager::init_storage_provider(const DoutPrefixProvider* dpp, CephContext* cct, const std::string svc, bool use_gc_thread, bool use_lc_thread, bool quota_threads, bool run_sync_thread, bool run_reshard_thread, bool use_cache, bool use_gc)
 {
-  if (svc.compare("rados") == 0) {
-    rgw::sal::Store* store = newStore();
-    RGWRados* rados = static_cast<rgw::sal::RadosStore* >(store)->getRados();
-
-    if ((*rados).set_use_cache(use_cache)
-                .set_use_datacache(false)
-                .set_use_gc(use_gc)
-                .set_run_gc_thread(use_gc_thread)
-                .set_run_lc_thread(use_lc_thread)
-                .set_run_quota_threads(quota_threads)
-                .set_run_sync_thread(run_sync_thread)
-                .set_run_reshard_thread(run_reshard_thread)
-                .initialize(cct, dpp) < 0) {
-      delete store;
-      return nullptr;
-    }
-    if (store->initialize(cct, dpp) < 0) {
-      delete store;
-      return nullptr;
-    }
-    return store;
-  }
-  else if (svc.compare("d3n") == 0) {
-    rgw::sal::RadosStore *store = new rgw::sal::RadosStore();
-    RGWRados* rados = new D3nRGWDataCache<RGWRados>;
-    store->setRados(rados);
-    rados->set_store(static_cast<rgw::sal::RadosStore* >(store));
-
-    if ((*rados).set_use_cache(use_cache)
-                .set_use_datacache(true)
-                .set_run_gc_thread(use_gc_thread)
-                .set_run_lc_thread(use_lc_thread)
-                .set_run_quota_threads(quota_threads)
-                .set_run_sync_thread(run_sync_thread)
-                .set_run_reshard_thread(run_reshard_thread)
-                .initialize(cct, dpp) < 0) {
-      delete store;
-      return nullptr;
-    }
-    return store;
+  const char *dlname = "/usr/lib64/ceph/librgw_sal_rados.so";
+  rgw::sal::Store* store = nullptr;
+  void *dl = nullptr;
+  rgw::sal::Store *(*newStore)(const DoutPrefixProvider *, CephContext *, bool, bool, bool, bool, bool, bool, bool, bool) = nullptr;
+  if (svc.compare("d3n") == 0) {
+    dlname = "/usr/lib64/ceph/librgw_sal_d3n.so";
```

Comment:
can we make this work without full paths here?

Reply:
Eventually I expect this to come from whatever the config solution is.  So yes.